php-checks/src/main/java/org/sonar/php/checks/LocalVariableShadowsClassFieldCheck.java

```diff
@@ -100,65 +113,74 @@ public void leaveFunctionScope() {
     }
   }
 
-  private boolean skip = false;
-
   @Override
-  public void init() {
-    subscribeTo(
-      PHPGrammar.CLASS_DECLARATION,
-      PHPGrammar.METHOD_DECLARATION,
-      PHPGrammar.FUNCTION_EXPRESSION,
-      PHPGrammar.ASSIGNMENT_EXPR,
-      PHPGrammar.ASSIGNMENT_BY_REFERENCE);
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    classState.clear();
+    super.visitCompilationUnit(tree);
   }
 
   @Override
-  public void visitFile(@Nullable AstNode astNode) {
+  public void visitClassDeclaration(ClassDeclarationTree tree) {
+    for (ClassMemberTree classMemberTree : tree.members()) {
+      if (classMemberTree.is(Kind.CLASS_PROPERTY_DECLARATION)) {
+        for (VariableDeclarationTree declaration : ((ClassPropertyDeclarationTree) classMemberTree).declarations()) {
+          classState.declareField(declaration.variableIdentifier().token());
+        }
+      }
+    }
+
+    classState.setClassName(tree);
+    super.visitClassDeclaration(tree);
     classState.clear();
-    skip = false;
   }
 
   @Override
-  public void visitNode(AstNode astNode) {
-    if (astNode.is(PHPGrammar.CLASS_DECLARATION)) {
-      declareClassField(astNode);
-      classState.setClassName(astNode);
-
-    } else if (classState.isInClass() && astNode.is(PHPGrammar.METHOD_DECLARATION)) {
-      if (isExcluded(astNode)) {
-        skip = true;
-      } else {
-        classState.newFunctionScope();
-        checkParameters(astNode);
-      }
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    if (!isExcluded(tree)) {
+      classState.newFunctionScope();
+      checkParameters(tree);
```

Comment:
@vilchik-elena what if a class property is declared after a method which uses a local variable with the same name?